plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java

```diff
@@ -350,6 +362,151 @@ public Optional<ProjectionApplicationResult<ConnectorTableHandle>> applyProjecti
                 precalculateStatisticsForPushdown));
     }
 
+    private Optional<ProjectionApplicationResult<ConnectorTableHandle>> applyProjectionExpressions(
+            ConnectorSession session,
+            JdbcTableHandle handle,
+            List<ConnectorExpression> projections,
+            Map<String, ColumnHandle> assignments)
+    {
+        int nextSyntheticColumnId = handle.getNextSyntheticColumnId();
+
+        ImmutableSet.Builder<JdbcColumnHandle> newColumnsBuilder = ImmutableSet.builder();
+        ImmutableList.Builder<Assignment> assignmentBuilder = ImmutableList.builder();
+        ImmutableMap.Builder<ConnectorExpression, Variable> newVariablesBuilder = ImmutableMap.builder();
+        ImmutableMap.Builder<String, ParameterizedExpression> columnExpressionsBuilder = ImmutableMap.builder();
+        Set<ConnectorExpression> translatedExpression = new HashSet<>();
+
+        if (isComplexExpressionPushdown(session)) {
+            for (ConnectorExpression projection : projections) {
+                RewrittenExpression rewrittenExpression = rewriteExpression(session, nextSyntheticColumnId, projection, assignments, translatedExpression);
+                nextSyntheticColumnId = rewrittenExpression.nextSyntheticColumnId;
+                newColumnsBuilder.addAll(rewrittenExpression.columnHandles);
+                assignmentBuilder.addAll(rewrittenExpression.assignments);
+                newVariablesBuilder.putAll(rewrittenExpression.syntheticVariables);
+                columnExpressionsBuilder.putAll(rewrittenExpression.columnExpressions);
+            }
+        }
+
+        List<JdbcColumnHandle> newColumns = ImmutableList.copyOf(newColumnsBuilder.build());
+
+        // Modify projections to refer to new variables
+        Map<ConnectorExpression, Variable> newVariables = newVariablesBuilder.buildOrThrow();
+        List<ConnectorExpression> newProjections = projections.stream()
+                .map(expression -> replaceWithNewVariables(expression, newVariables))
+                .collect(toImmutableList());
+
+        List<Assignment> outputAssignments = assignmentBuilder.build();
+
+        if (!newVariables.isEmpty()) {
```

Comment:
this will disappear after 
https://github.com/trinodb/trino/pull/22203/files#r1630996112